README.md
# Mathematical Operation Visualizer 
[TOC]

## Purpose

The main purpose, or goal of this program is that it allows the users to input a mathematical operation, and it would express that operations in the human readable form. This program allows the user to input variety of operations, and also add complex operations to them which is then visualized. Also the program allows you to customize the output, with different background colors and different colors for the fonts. Along with that the program allows user to download the output, as a PNG and save it on their device. 

## How to Run the Program
Open the terminal, and enter the directory of where the program is. Then enter the following code:	

```bash
$$ javac Main.java
$$ Main.java
```

Then a GUI should open where you can use it.

## How it Works

Once the program is started, it would then show a GUI where it contains certain elements which the user could utilized to get its tasks done. 

### Features in GUI:

#### Input Area

- **Input Field:** In this field, the user can input the mathematical operation, which the user wished to get visualized. 
- **Evaluate Button:** When clicked on this button, whatever is inside the *input area* would be taken and then evaluated, when would then be shown in the output field

#### Output Area

- It is an **open** and **large** space at which the human readable form of the mathematical operations would be shown. 

#### Menu Bar

- **File:** It contains two buttons. One is *<u>Download Image</u>* which would download whatever is inside the *input field* and download it as PNG. Another button is *<u>Help</u>* which would open another windows, at which there would be few inputs which could help the user cruise through the program.
- **Edit:** This contains sever options, from which could be used to customize the *input field*. The background color, or the font color could be changed depending on the user's preferences. 

## Instructions

In order for to user to use the program properly, there are indeed few things which the user has to be careful of. Firstly is, that this program does not support decimal numbers. This program could only represent whole number (both positive and negative numbers). If some by chance the program crashes, or outputs an error then simply restart the program. Another key point, to take care of is that there is limited amount of operations which are supported. Operators such as "=" aren't supported, and would lead to the program crashing. 

## Supported Operators

Bellow is the list of all the supported operator which the program can represent (examples bellow are the inputs):

1. **Literals** (eg. *1*)
2. **Variables** (eg. *x*)
3. **Addition** (eg. *1+1*)
4. **Subtraction** (eg. *1-1*)
5. **Multiplication** (eg. *1*1*)
6. **Division** (eg. *1/1*)
7. **Modulus** (eg. *1%1*)
8. **Greater** (eg. *1<2*)
9. **Smaller** (eg. *2>1*)
10. **Exponent** (eg. *1^2*)
11. **sin** (eg. *sin(x)*)
12. **cos** (eg. *cos(x)*)
13. **tan** (eg. *tan(x)*)
14. **ln** (eg. *ln(x))*
15. **inverse** (eg. *sin^-1(2)*)

## PMD and CPD Violations

From the mvn check, there are indeed some Violations. One of those violations which repeats a lot is the violations of duplication. This violations exists, because many times same code is used to style the buttons, or to just have systematic structure in the code. For instance in the Help method there tends to be a lot of duplication of code as because there had to be a systemic order, so that it easy to read and follow. This problem could be avoided if sub-classes are made where  such code is actually called upon, rather then being repeating. This would improve the performance and would make it much more efficient as well.

## Future Scope

There is indeed a lot of scope for this program. In future for instance the division line could be actually made horizontal. Also along with that the actuall calculation could be done for the program, where it also shows the result of the input. It would have been realized that a class called as *draw* is actually not that usefully. However, that class is kept for the improvements in the future, as this class could be used to implement those features. Along with that more customization would also be appreciated as well in the future updated. 


CHANGES.md
# Changes

## Milestone 1: May 7
* Initialized repo
* Created README.md
* Created CHANGES.md
* Imported code from Labs 8 and 9

## Milestone 2: May 14
* Added code from Lab 10
* Modified the code, so that it returns a string except of a node.

## Milestone 3: May 24
* Added extra features (different operations are now possible)
* Also it prints out the visual repersentation of those expression in the output field

## Milestone 4: May 30

- Reworked on the logic of visualizing the expression. Made it more efficition by not calling *repaint()* all the time.
- Added more opertors, and now also division and power is supported as well.

## Deadline: June 4

- Added the functionality to download the output as an image.
- Added the functionlity to customize the output with different colors and font color/
- Added an extra window which provides support to the user.


DrawFrame.java
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.image.BufferedImage; 
import java.io.File;
import javax.imageio.ImageIO;
import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JMenu; 
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JPanel;
import javax.swing.JTextField;
import javax.swing.SwingConstants;





/**
 * The GUI of the program, where all the object are pressent. For instance the,
 * menu, buttons, panel and label.
 *
 * @author Harkeerat Singh Sawhney
 * @version 3.06.2021
 */
public class DrawFrame extends JFrame
{
    //Label object where the final output is shown.
    private JLabel label;
    
    //MenuBar object where the MenuBar is made.
    private final JMenuBar menuBar;
    //Menuobject where the menu's needed are made.
    private final JMenu file;
    private final JMenu edit;
    private final JMenu backgroundColor; 
    private final JMenu fontColor;
    
    //Custom RGB colors.
    private final Color blue = new Color(51,153,255);    //Light Blue Color.
    private final Color yellow = new Color(226,249,80);  //Light Yellow Color.
    private final Color green = new Color(80,249,165);   //Light Green Color.
    private final Color pink = new Color(185,140,238);   //Light Pink Color.
    private final Color red = new Color(255,63,114);     //Light Red Color.
    
    /**
     * Constructor for objects of class DrawFrame.
     * @param draw which is used to get the function
     */
    public DrawFrame(final Draw draw)
    {
        super();
        setTitle("Math Operation Visualizer");                 //Title of the Program.
        setLayout(new BorderLayout());
        this.setPreferredSize(new Dimension(800, 500));        //Size of the Resolution.
        final JPanel panel = new JPanel();
        
        //Making new Objects for the MenuBar
        menuBar = new JMenuBar();
        file = new JMenu("File");      
        edit = new JMenu("Edit");
        backgroundColor = new JMenu("Background Color");
        fontColor = new JMenu("Font Color");
        
        //Buttons for the change of color of the background.
        final JButton blueBackground = new JButton("Blue");
        blueBackground.setPreferredSize(new Dimension(90, 20));
        blueBackground.setBackground(Color.WHITE);
        
        final JButton yellowBackground = new JButton("Yellow");
        yellowBackground.setPreferredSize(new Dimension(90, 20));
        yellowBackground.setBackground(Color.WHITE);
        
        final JButton greenBackground = new JButton("Green");
        greenBackground.setPreferredSize(new Dimension(90, 20));
        greenBackground.setBackground(Color.WHITE);
        
        final JButton pinkBackground = new JButton("Pink");
        pinkBackground.setPreferredSize(new Dimension(90, 20));
        pinkBackground.setBackground(Color.WHITE);
        
        final JButton redBackground = new JButton("Red");
        redBackground.setPreferredSize(new Dimension(90, 20));
        redBackground.setBackground(Color.WHITE);     
        
        //Buttons for the change color of the text.
        final JButton blackFont = new JButton("Black");
        blackFont.setPreferredSize(new Dimension(90, 20));
        blackFont.setBackground(Color.WHITE);   
        
        final JButton redFont = new JButton("Red");
        redFont.setPreferredSize(new Dimension(90, 20));
        redFont.setBackground(Color.WHITE);   
        
        final JButton blueFont = new JButton("Blue");
        blueFont.setPreferredSize(new Dimension(90, 20));
        blueFont.setBackground(Color.WHITE);   
        
        final JButton yellowFont = new JButton("Yellow");
        yellowFont.setPreferredSize(new Dimension(90, 20));
        yellowFont.setBackground(Color.WHITE);   
        
        final JButton whiteFont = new JButton("White");
        whiteFont.setPreferredSize(new Dimension(90, 20));
        whiteFont.setBackground(Color.WHITE);   
                
        //Button for the download the lable as PNG
        final JButton download = new JButton("Download Image");  
        download.setPreferredSize(new Dimension(150, 30));
        download.setBackground(Color.WHITE);
        
        final JButton help = new JButton("Help");  
        help.setPreferredSize(new Dimension(150, 30));
        help.setBackground(Color.WHITE);     
        
        //Addiing the object to the menubar
        menuBar.add(file);
        menuBar.add(edit);
        edit.add(backgroundColor);
        edit.add(fontColor);
        file.add(download);
        file.add(help);
        
        //Adding the color's buttons.
        backgroundColor.add(blueBackground);
        backgroundColor.add(yellowBackground);
        backgroundColor.add(greenBackground);
        backgroundColor.add(pinkBackground);
        backgroundColor.add(redBackground);
        
        //Adding the font's buttons.
        fontColor.add(blackFont);
        fontColor.add(redFont);
        fontColor.add(blueFont);
        fontColor.add(yellowFont);
        fontColor.add(whiteFont);
        
        //Adding the menubar to the frame.
        add(menuBar, BorderLayout.NORTH);
        setJMenuBar(menuBar);   
        
        //Making the button, which would evalute the expression in the field.
        final JButton ev = new JButton("Evaluate");
        ev.setPreferredSize(new Dimension(100, 30));
        
        //The text field where user's inputs the expression.
        final JTextField expressionField = new JTextField(draw.getFunction().getExpression(), 15);
        Font bigFont = expressionField.getFont().deriveFont(Font.PLAIN, 30f);
        expressionField.setFont(bigFont);        
        
        panel.add(expressionField);
        panel.add(ev);
        add(panel, BorderLayout.NORTH);
        
        //Label,where the expression is shown to the users.
        label = new JLabel(draw.getFunction().getExpression(), SwingConstants.CENTER);
        label.setBounds(50, 50, 100, 30);
        label.setFont(new Font("Verdana", Font.PLAIN, 50));
        label.setBorder(BorderFactory.createLineBorder(Color.black));
        label.setBackground(Color.WHITE);
        label.setOpaque(true);
        
        
        add(label, BorderLayout.CENTER);
        
        ev.addActionListener(new ActionListener() { 
            public void actionPerformed(final ActionEvent e) {
                final String text = expressionField.getText();
                draw.getFunction().setExpression(text);
                final String newText = draw.getFunction().getExpression();
                label.setText(newText);
            }
        });
        
        yellowBackground.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setBackground(yellow);
                label.setOpaque(true);
            }
        });
        
        greenBackground.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setBackground(green);
                label.setOpaque(true);
            }
        });
        
        pinkBackground.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setBackground(pink);
                label.setOpaque(true);
            }
        });
        
        redBackground.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setBackground(red);
                label.setOpaque(true);
            }
        });
        
        blueBackground.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setBackground(blue);
                label.setOpaque(true);
            }
        });
        
        
        blackFont.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setForeground(Color.black);
                label.setOpaque(true);
            }
        });
        
        redFont.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setForeground(Color.red);
                label.setOpaque(true);
            }
        });
        
        yellowFont.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setForeground(Color.yellow);
                label.setOpaque(true);
            }
        });
        
        blueFont.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setForeground(Color.blue);
                label.setOpaque(true);
            }
        });
        
        whiteFont.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                label.setForeground(Color.white);
                label.setOpaque(true);
            }
        });
        
        
        download.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                BufferedImage img = new BufferedImage(label.getWidth(), label.getHeight(), BufferedImage.TYPE_INT_ARGB);
                Graphics2D g2d = img.createGraphics();
                label.printAll(g2d);
                g2d.dispose();
                try {
                    ImageIO.write(img, "png", new File("C:\\Users\\Admin\\Desktop\\Math_Expresion.png"));
                } catch (java.io.IOException ioe) {
                    ioe.printStackTrace();
                }
            }
        });
        
        help.addActionListener(new ActionListener() {
            public void actionPerformed(final ActionEvent e) {
                HelpPage.main();
            }
        });
        
        pack();
    }

}
